azavea/franklin
# Add transaction endpoints

## Overview

This PR adds transaction endpoints and a `--with-transactions` flag to server startup to turn them on if you want them.

### Checklist

- ~New tests have been added or existing tests have been modified~ we need generators from `stac4s` for testing these (or to copy-paste like 20 gens + arbitraries) -- opened azavea/raster-foundry-platform#938 instead

### Notes

The flag controls the presence of the endpoints in actual routing and in docs.

### Testing Instructions

- do you have the Panama data? If not, wave a magic wand and load the Panama data
- get yourself a nice json file with an item in it: `http :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/890b0f0d-2b6c-47fa-82e0-980212faf557 | jq . > item.json`
- change its id -- I incremented the last 7 to an 8 -- and POST it to the collection items service: `cat item.json | http :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/`. Note the ETag in the response and save it as `$ETag` in your shell
- add a property `foo: bar` in the json file
- try to put it to the item: `cat item.json | http PUT :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/890b0f0d-2b6c-47fa-82e0-980212faf558` -- you should get a `BadRequest` because the ETag is missing
- try to put it to the item with a bogus ETag: `cat item.json | http PUT :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/890b0f0d-2b6c-47fa-82e0-980212faf558 If-Match:abcde` -- you should get a 412 beacuse the ETag doesn't match
- try to put it to the item with the correct ETag: `cat item.json | http PUT :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/890b0f0d-2b6c-47fa-82e0-980212faf558 If-Match:$ETag` -- it should succeed. Note the new ETag and again save it as `$ETag`
- try to patch the item with a bad json format, e.g. set `properties` to something that isn't an object: `echo '{"properties": {"label:tasks": ["lol"]}}' | http PATCH :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/890b0f0d-2b6c-47fa-82e0-980212faf558 If-Match:$ETag` -- it should fail with a pretty helpful message
- try to patch the item with a good json format: `echo '{"properties": {"label:tasks": ["lol"]}}' | http PATCH :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/890b0f0d-2b6c-47fa-82e0-980212faf558 If-Match:$ETag` ~**NOTE: BROKEN. I don't know why. Json merge isn't cooperating for object updates.**~ https://github.com/azavea/franklin/pull/65/commits/cfd7831a558f741a8f999e787a35c6c9032a8e24 oops
- delete your item (note -- doesn't require an `If-Match` header): `http DELETE :9090/collections/0be4882c-a509-43f4-81c9-ad60bdada264/items/890b0f0d-2b6c-47fa-82e0-980212faf558`
- get the api spec -- `http :9090/open-api/spec.yaml` -- and look at the beautiful, POST, PUT, PATCH, and DELETE endpoints
- stop your server, and restart it without `--with-transactions`
- try to send any sort of POST, PUT, PATCH, or DELETE request to items
- it should 404
- look at the api spec and note that the transaction endpoints are gone

Closes #4
